node/lib/util/stitch_util.js

```diff
@@ -0,0 +1,637 @@
+ * specified `sha` has beren converted.
+ *
+ * @param {String} sha
+ * @return {String}
+ */
+exports.convertedRefName = function (sha) {
+    return "refs/stitched/converted/" + exports.splitSha(sha);
+};
+
+/**
+ * Return the name of the ref indicating that the submodule commit having the
+ * specified `subSha` has been fetched for the specified `metaSha`; this ref
+ * also serves to root that submodule commit.
+ *
+ * @param {String} metaSha
+ * @param {String} subSha
+ * @return {String}
+ */
+exports.fetchedSubRefName = function (metaSha, subSha) {
+    return "refs/stitched/fetched/" + exports.splitSha(metaSha) + "/sub/" +
+        exports.splitSha(subSha);
+};
+
+/**
+ * Return a summary for the specified `commit` for the submodule having the
+ * specifiewd `name` to be used to describe changes that are rolled into a
+ * meta-repo commit.
+ *
+ * @param {String} name
+ * @param {Commit} subCommit
+ */
+exports.summarizeSubCommit = function (name, subCommit) {
+    assert.isString(name);
+    assert.instanceOf(subCommit, NodeGit.Commit);
+
+    const author = subCommit.author();
+    return `\
+Includes changes from submodule ${name} on ${subCommit.id().tostrS()}.
+Author: ${author.name()} <${author.email()}>
+Date:   ${Commit.formatCommitTime(author.when())}
+
+${subCommit.message()}
+`;
+};
+
+/**
+ * Return true if there is no difference in the specified `newTree` and
+ * `originalTree` represent the same value.  If `newTree` represents the same
+ * value as `originalTree` if they have the same id, or if
+ * `null === originalTree` and `newTree` has no entries.
+ *
+ * @param {NodeGit.Tree}      newTree
+ * @param {NodeGit.Tree|null} originalTree
+ * @return {Bool}
+ */
+exports.isEmpty = function (newTree, originalTree) {
+    assert.instanceOf(newTree, NodeGit.Tree);
+    if (null !== originalTree) {
+        assert.instanceOf(originalTree, NodeGit.Tree);
+    }
+    if (null === originalTree) {
+        return 0 === newTree.entries().length;
+    }
+    return newTree.id().tostrS() === originalTree.id().tostrS();
+};
+
+/**
+ * From a map containing a shas mapped to sets of direct parents, and the
+ * specified starting `entry` sha, return a list of all shas ordered from least
+ * to most dependent, that is, no sha will appear in the list before any of its
+ * ancestors.  If no relation exists between two shas, they will be ordered
+ * alphabetically.  Note that it is valid for a sha to exist as a parent from a
+ * sha in `parentMap`, however, the behavior is undefined if there are entries
+ * in 'parentMap' that are not reachable from 'entry'.
+ *
+ * @param {String} entry
+ * @param {Object} parentMap  from sha to Set of its direct parents
+ * @return {[String]}
+ */
+exports.listCommitsInOrder = function (entry, parentMap) {
+    assert.isString(entry);
+    assert.isObject(parentMap);
+
+    // First, compute the levels of the commits.  A level '0' means that a
```

Comment:
FWIW, git internally uses the term "generation" for this (in some new code that I was looking at recently).

And I would say "a generation N means that the max generation of the commit's parents is (N - 1)", but maybe that's not actually clearer.

Reply:
I'll use the git term (In the past I've used the term "levalization")